File: sdk/openfeature/examples/contract_smoke.mjs
/**
 * Contract smoke for the OpenFeature (JS) provider.
 *
 * Run from the repo root:
 *   cd sdk/openfeature && npm run build --silent && EXPERIMENTLY_API_KEY=... node examples/contract_smoke.mjs
 *
 * Flags go through the OpenFeature API; experiment assignment and tracking go
 * through `provider.client` (the underlying JS SDK client).
 * Prints exactly one JSON line on stdout and exits 0; on failure one line to stderr and exit 1.
 */
import { randomUUID } from 'node:crypto';
import { OpenFeature } from '@openfeature/server-sdk';
import pkg from '../dist/index.js';

const { ExperimentationProvider, ExperimentationClient } = pkg;

const apiUrl = process.env.EXPERIMENTLY_API_URL || 'http://localhost:8000';
const apiKey = process.env.EXPERIMENTLY_API_KEY;
const experimentKey = process.env.CONTRACT_EXPERIMENT_KEY || 'sdk_contract_ab';
const flagKey = process.env.CONTRACT_FLAG_KEY || 'sdk_contract_flag';
const userId = process.env.CONTRACT_USER_ID || `smoke-${randomUUID()}`;

function fail(message) {
  process.stderr.write(`openfeature contract smoke failed: ${message}\n`);
  process.exit(1);
}

if (!apiKey) fail('EXPERIMENTLY_API_KEY is required');

const trackErrors = [];
const sdk = new ExperimentationClient({
  apiUrl,
  apiKey,
  onError: (error, operation) => trackErrors.push(`${operation}: ${error.message}`),
});
const provider = new ExperimentationProvider({ apiKey, baseUrl: apiUrl, client: sdk });

try {
  await OpenFeature.setProviderAndWait(provider);
  const flags = OpenFeature.getClient('contract-smoke');

  // 1. Assign twice; clearCache() in between makes the second call hit the server.
  const first = await sdk.getAssignment(experimentKey, { userId, attributes: { source: 'contract-smoke' } });
  sdk.clearCache();
  const second = await sdk.getAssignment(experimentKey, { userId });
  if (!['control', 'treatment'].includes(first.variantName)) {
    fail(`unexpected variant_name ${JSON.stringify(first.variantName)}`);
  }
  const sticky = first.variantName === second.variantName && first.variantId === second.variantId;
  if (!sticky) fail(`assignment not sticky: ${first.variantName} then ${second.variantName}`);

  // 2. Evaluate the flag through OpenFeature.
  const details = await flags.getBooleanDetails(flagKey, false, { targetingKey: userId });
  if (details.errorCode) fail(`flag: ${details.errorCode} ${details.errorMessage ?? ''}`);
  if (typeof details.value !== 'boolean') fail('flag value is not a boolean');

  // 3. Track with an explicit experiment key.
  await sdk.track(userId, 'purchase', { value: 12.5, experimentKey });
  if (trackErrors.length) fail(`track: ${trackErrors[0]}`);

  // 4. Track without a key: fans out to the cached assignment + the flag evaluated via OpenFeature.
  if (sdk.getAssignments(userId).length !== 1 || sdk.getEvaluatedFlags(userId).length !== 1) {
    fail('cache does not hold one assignment and one flag; fan-out would send nothing');
  }
  await sdk.track(userId, 'page_view', { properties: { page: '/contract-smoke' } });
  if (trackErrors.length) fail(`fan-out: ${trackErrors[0]}`);

  // 4b. Explicit two-event batch.
  const batch = await sdk.trackBatch([
    { userId, eventName: 'add_to_cart', experimentKey, value: 1 },
    { userId, eventName: 'flag_seen', featureFlagKey: flagKey },
  ]);
  if (batch.failureCount > 0 || batch.successCount !== 2) fail(`batch: ${JSON.stringify(batch)}`);

  process.stdout.write(
    JSON.stringify({
      sdk: 'openfeature',
      assign: { variant_name: second.variantName, is_control: second.isControl, sticky },
      flag: { enabled: details.value },
      track: { ok: true },
      fanout: { ok: true },
    }) + '\n'
  );
  await OpenFeature.close();
} catch (err) {
  fail(err && err.message ? err.message : String(err));
}

File: sdk/openfeature-python/examples/contract_smoke.py
#!/usr/bin/env python3
"""Contract smoke for the OpenFeature Python provider against a live backend.

Run from the repository root (no install needed; ``openfeature-sdk`` must be importable):

    EXPERIMENTLY_API_KEY=<key> python sdk/openfeature-python/examples/contract_smoke.py

Env: EXPERIMENTLY_API_URL (default http://localhost:8000), EXPERIMENTLY_API_KEY (required),
CONTRACT_EXPERIMENT_KEY (default sdk_contract_ab), CONTRACT_FLAG_KEY (default sdk_contract_flag),
CONTRACT_USER_ID (default: random "smoke-<uuid>").

The flag is resolved through the OpenFeature API (boolean resolution); experiment assignment
and tracking — which OpenFeature does not model — go through the provider's underlying
``experimentation`` client. Prints exactly one JSON line on stdout and exits 0; on failure
prints one line to stderr and exits 1.
"""

from __future__ import annotations

import json
import os
import sys
import uuid
from pathlib import Path

_HERE = Path(__file__).resolve()
for _path in (_HERE.parents[1] / "src", _HERE.parents[2] / "python"):
    sys.path.insert(0, str(_path))

from openfeature import api  # noqa: E402
from openfeature.evaluation_context import EvaluationContext  # noqa: E402

from experimentation import ExperimentationError  # noqa: E402
from experimentation_openfeature import ExperimentationProvider  # noqa: E402


def fail(message: str) -> None:
    print(message, file=sys.stderr)
    sys.exit(1)


def main() -> None:
    api_url = os.environ.get("EXPERIMENTLY_API_URL", "http://localhost:8000")
    api_key = os.environ.get("EXPERIMENTLY_API_KEY")
    if not api_key:
        fail("EXPERIMENTLY_API_KEY is required")
    experiment_key = os.environ.get("CONTRACT_EXPERIMENT_KEY", "sdk_contract_ab")
    flag_key = os.environ.get("CONTRACT_FLAG_KEY", "sdk_contract_flag")
    user_id = os.environ.get("CONTRACT_USER_ID") or f"smoke-{uuid.uuid4()}"
    attributes = {"source": "contract_smoke", "sdk": "openfeature-python"}

    provider = ExperimentationProvider(api_key=api_key, base_url=api_url)
    api.set_provider(provider)
    of_client = api.get_client()
    sdk = provider.client

    # 1. Assign twice via the underlying SDK client (cache cleared in between so the
    #    second call really hits the server and proves stickiness).
    try:
        first = sdk.get_assignment(experiment_key, user_id, attributes)
        sdk.clear_cache()
        second = sdk.get_assignment(experiment_key, user_id, attributes)
    except ExperimentationError as exc:
        fail(f"assign failed: {exc} (status={exc.status}, body={exc.body!r})")
    if first.variant_name not in ("control", "treatment"):
        fail(f"unexpected variant_name {first.variant_name!r}")
    sticky = first == second
    if not sticky:
        fail(f"assignment not sticky: {first.variant_name} then {second.variant_name}")

    # 2. Evaluate the flag through OpenFeature (boolean resolution).
    details = of_client.get_boolean_details(
        flag_key, False, EvaluationContext(targeting_key=user_id, attributes=attributes)
    )
    if details.error_code is not None:
        fail(f"flag evaluation failed: {details.error_code} {details.error_message}")
    if not isinstance(details.value, bool):
        fail(f"flag value is not a bool: {details.value!r}")

    # 3. Track with an experiment key -> POST /tracking/track.
    if not sdk.track(user_id, "purchase", event_value=12.5, experiment_key=experiment_key):
        fail("track(purchase) with experiment_key was not accepted")

    # 4. Track without a key -> fans out to the cached assignment + flag, then a 2-event batch.
    if not sdk.track(user_id, "page_view", properties={"page": "/smoke"}):
        fail("track(page_view) without a key was not accepted")
    batch = sdk.track_batch(
        [
            {"event_name": "page_view", "user_id": user_id, "experiment_key": experiment_key},
            {"event_name": "click", "user_id": user_id, "feature_flag_key": flag_key},
        ]
    )
    if not batch.ok:
        fail(f"track_batch failed: {batch.errors}")

    print(
        json.dumps(
            {
                "sdk": "openfeature-python",
                "assign": {
                    "variant_name": first.variant_name,
                    "is_control": first.is_control,
                    "sticky": sticky,
                },
                "flag": {"enabled": details.value},
                "track": {"ok": True},
                "fanout": {"ok": True},
            },
            separators=(",", ":"),
        )
    )


if __name__ == "__main__":
    try:
        main()
    except SystemExit:
        raise
    except Exception as exc:  # noqa: BLE001 - one line on stderr, exit 1
        fail(f"{type(exc).__name__}: {exc}")
